File: src/test/java/chat/teco/tecochat/auth/domain/AuthenticatorTest.java
package chat.teco.tecochat.auth.domain;

import static chat.teco.tecochat.auth.exception.AuthenticationExceptionType.NO_REGISTERED_MEMBER;
import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.mock;

import chat.teco.tecochat.auth.exception.AuthenticationException;
import chat.teco.tecochat.common.exception.BaseExceptionType;
import chat.teco.tecochat.member.domain.MemberRepository;
import chat.teco.tecochat.member.fixture.MemberFixture.말랑;
import java.util.Optional;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.DisplayNameGeneration;
import org.junit.jupiter.api.DisplayNameGenerator.ReplaceUnderscores;
import org.junit.jupiter.api.Test;

@SuppressWarnings("NonAsciiCharacters")
@DisplayNameGeneration(ReplaceUnderscores.class)
@DisplayName("Authenticator 은(는)")
class AuthenticatorTest {

    private final MemberRepository memberRepository = mock(MemberRepository.class);
    private final Authenticator authenticator = new Authenticator(memberRepository);

    @Test
    void Base64로_인코딩된_이름으로_인증() {
        // given
        given(memberRepository.findByName("mallang")).willReturn(Optional.of(말랑.회원()));

        // when
        Long id = authenticator.authenticateWithBase64("bWFsbGFuZw==").id();

        // then
        assertThat(id).isEqualTo(말랑.ID);
    }

    @Test
    void 가입한_회원이_없으면_오류() {
        // given
        given(memberRepository.findByName("mallang")).willReturn(Optional.empty());

        // when
        BaseExceptionType baseExceptionType = assertThrows(AuthenticationException.class, () ->
                authenticator.authenticateWithBase64("bWFsbGFuZw==")
        ).exceptionType();

        // then
        assertThat(baseExceptionType).isEqualTo(NO_REGISTERED_MEMBER);
    }
}

File: src/test/kotlin/domain/AuthenticatorTest.kt
package chat.teco.tecochat.auth.domain

import chat.teco.tecochat.createMember
import chat.teco.tecochat.member.domain.MemberRepository
import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.shouldBe
import io.mockk.every
import io.mockk.mockk
import org.junit.jupiter.api.assertThrows
import java.util.*

class AuthenticatorTest : StringSpec({
    val memberRepository = mockk<MemberRepository>()

    val authenticator = Authenticator(memberRepository)

    "Base64로 인코딩된 닉네임으로 인증한다" {
        val member = createMember()
        every { memberRepository.findByName(any()) } returns Optional.of(member)

        val result = authenticator.authenticateWithBase64("bWFsbGFuZw==")

        result shouldBe member
    }

    "닉네임에 해당하는 회원이 존재하지 않는다면 예외를 던진다" {
        every { memberRepository.findByName(any()) } returns Optional.empty()

        assertThrows<IllegalArgumentException> {
            authenticator.authenticateWithBase64("mallang")
        }
    }
})
